inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java

```diff
@@ -0,0 +1,362 @@
+            }
+
+            LOG.info("Successfully restore the offsets {}.", currentOffsets);
+        } else {
+            LOG.info("No restore offsets.");
+        }
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        ConsumerConfig consumerConfig =
+                new ConsumerConfig(masterAddress, consumerGroup);
+        consumerConfig.setConsumePosition(consumeFromMax
+                ? ConsumePosition.CONSUMER_FROM_MAX_OFFSET_ALWAYS
+                : ConsumePosition.CONSUMER_FROM_FIRST_OFFSET);
+
+        consumerConfig.setMsgNotFoundWaitPeriodMs(messageNotFoundWaitPeriod.toMillis());
+
+        final int numTasks = getRuntimeContext().getNumberOfParallelSubtasks();
+
+        messageSessionFactory = new TubeSingleSessionFactory(consumerConfig);
+
+        messagePullConsumer =
+                messageSessionFactory.createPullConsumer(consumerConfig);
+        messagePullConsumer
+                .subscribe(topic, tidSet);
+        messagePullConsumer
+                .completeSubscribe(sessionKey, numTasks, true, currentOffsets);
+
+        running = true;
+    }
+
+    @Override
+    public void run(SourceContext<T> ctx) throws Exception {
+
+        Instant lastConsumeInstant = Instant.now();
+
+        while (running) {
+
+            ConsumerResult consumeResult = messagePullConsumer.getMessage();
+            if (!consumeResult.isSuccess()) {
+                if (!(consumeResult.getErrCode() == 400
+                        || consumeResult.getErrCode() == 404
+                        || consumeResult.getErrCode() == 405
+                        || consumeResult.getErrCode() == 406
+                        || consumeResult.getErrCode() == 407
+                        || consumeResult.getErrCode() == 408)) {
+                    LOG.info("Could not consume messages from tubemq (errcode: {}, "
+                                    + "errmsg: {}).", consumeResult.getErrCode(),
+                            consumeResult.getErrMsg());
+                }
+
+                Duration idleTime =
+                        Duration.between(lastConsumeInstant, Instant.now());
+                if (idleTime.compareTo(maxIdleTime) > 0) {
+                    ctx.markAsTemporarilyIdle();
+                }
+
+                continue;
+            }
+
+            List<Message> messageList = consumeResult.getMessageList();
+            lastConsumeInstant = Instant.now();
+
+            List<T> records = new ArrayList<>();
+            if (messageList != null) {
+                lastConsumeInstant = Instant.now();
+
+                for (Message message : messageList) {
+                    T record =
+                            deserializationSchema.deserialize(message.getData());
+                    records.add(record);
+                }
+            }
+
+            synchronized (ctx.getCheckpointLock()) {
+
+                for (T record : records) {
+                    ctx.collect(record);
```

Comment:
can checkpoint lock only lock currentoffsets ? this lock is used in several places which may cause degration in term s of performance

Reply:
https://github.com/apache/inlong/pull/4891#issuecomment-1177005578

Yes, you are right.
This is going to be an improvement term.Base on @thesumery advice.